cloudferrylib/utils/retrying.py

```diff
@@ -0,0 +1,197 @@
+        self.predicate = predicate
+        self.predicate_args = predicate_args
+        self.predicate_kwargs = predicate_kwargs
+        self.predicate_retval_as_arg = predicate_retval_as_arg
+        self.retry_message = retry_message
+        self.stop_on_return_value = stop_on_return_value
+
+    def max_attempts_reached(self):
+        return self.attempt >= self.max_attempts
+
+    def increment_attempts(self):
+        self.attempt += 1
+
+    def timedout(self):
+        return self.total_time >= self.max_time
+
+    def update_total_time(self):
+        if self.wait_exponential:
+            self.timeout *= 2
+
+        self.total_time += self.timeout
+
+    def sleep(self):
+        time.sleep(self.timeout)
+
+    def get_retry_message(self, auto_message):
+        if self.retry_message:
+            return self.retry_message
+        else:
+            return auto_message
+
+    def handle_predicate(self, retval):
+        if self.predicate and callable(self.predicate):
+            if self.predicate_retval_as_arg:
+                args = [retval]
+            else:
+                args = self.predicate_args or []
+            kwargs = self.predicate_kwargs or {}
+            if not self.predicate(*args, **kwargs):
+                default_msg = "Function exited with '{}'".format(retval)
+                msg = self.get_retry_message(default_msg)
+                raise RetryRequired(msg)
+
+    def handle_return_on_value(self, retval):
+        if self.stop_on_return_value and retval != self.return_value:
+            msg = "Return value '{}' didn't match expected '{}'"
+            raise RetryRequired(msg.format(retval, self.return_value))
+
+        if self.retry_on_return_value and retval == self.return_value:
+            msg = self.get_retry_message(
+                "Unexpected return value: {}".format(retval))
+            raise RetryRequired(msg)
+
+    def run(self, func, *args, **kwargs):
+        stop_retrying = self.max_attempts_reached
+        step_action = self.increment_attempts
+
+        if self.max_time is not UNLIMITED:
+            stop_retrying = self.timedout
+            step_action = self.update_total_time
+
+        retval = None
+        failing = True
+        last_exception = None
+
+        while failing and not stop_retrying():
+            step_action()
+            try:
+                retval = func(*args, **kwargs)
+
+                self.handle_predicate(retval)
+                self.handle_return_on_value(retval)
+
+                failing = False
+            except Exception as e:
+                if any([isinstance(e, expected)
+                        for expected in self.expected_exceptions]):
+                    raise e
+                last_exception = e
+                LOG.debug("%s, retrying", e)
+                self.sleep()
+
+        if self.raise_error and failing:
+            if self.reraise_original_exception and last_exception is not None:
+                raise last_exception
```

Comment:
Sergey, isn't `raise last_exception[0], last_exception[1], last_exception[2]` equivalent to `raise last_exception`?